Update CONTRIBUTING.md (#1576)
holochain/holochain: CONTRIBUTING.md

## Git Hygiene
This section describes our practices and guidelines for using git and making changes to the repo.

* We use Github's pull requests as our code review tool
* We encourage any dev to comment on pull requests and we think of the pull request not as a "please approve my code" but as a space for co-developing, i.e. asynchronous "pair-coding" of a sort.
* We develop features on separate branches
* We use merge (not rebase) so that commits related to a ticket can be retroactively explored.
* In most repos development happens on a `develop` branch which gets merged to `main` when there's a release.

### Pull requests

We warmly welcome pull requests for bug fixes, bug reproductions, documentation improvements, and any other "obviously good" enhancements to the codebase. If you are unsure if an enhancement is "obviously good", please coordinate with us first through a GitHub issue, or through our forums or Discord channel. We reserve the right to close any PR which doesn't fit our overall development trajectory, but we will gladly review any PRs and work with authors who have taken the time to identify a real problem or need and take steps to address it.

> To open a PR, fork our Github repo, create a branch whose name describes your fix, and base your pull request on our `develop` branch.

Reviewer: ```suggestion
To open a PR, fork our Github repo, create a branch whose name describes your fix, and base and target your pull request to our `develop` branch.
```

> If you add or change functionality, be sure to add both unit tests and integration tests to show that it works. Pull requests without tests will most likely not be accepted!

Reviewer: ```suggestion
If you add or change functionality, be sure to add or update both unit tests and integration tests to show that it works. Pull requests without tests compromise the code quality and can not be accepted.
```

## Bug reports

> The simplest way to report a bug is via [Github Issues](https://github.com/holochain/holochain/issues/new/choose) by selecting the Bug Report issue type. Please fill out all relevant areas of the bug report, including steps to reproduce. If your report is not specific enough, we will have a hard time addressing it without further followup.

Reviewer: ```suggestion
The simplest way to report a bug is via [Github Issues](https://github.com/holochain/holochain/issues/new/choose) by selecting the Bug Report issue type. Please fill out all relevant areas of the bug report, including steps to reproduce. If your report is not specific enough, we will have a hard time addressing it without further follow-up.
```

## Bug fixes and minimal reproductions

We gladly welcome pull requests that help us identify and fix bugs!

The end goal of addressing any bug is to have a test written in our codebase to reproduce the bug, and of course to implement the fix for the bug. A PR with at least a minimal reproduction demonstrating the bug is extremely helpful, even if the fix has not been discovered.

To write a minimal reproduction of a problem discovered "in the wild", we recommend you to write a sample zome, DNA, or hApp which demonstrates the problem, and open a PR with your failing test. We have a library called [`sweettest`](https://docs.rs/holochain/latest/holochain/sweettest/index.html) which is well-suited to the task of testing the behavior of Holochain applications. 

When writing your reproduction PR, you can recreate the problematic part of your app in one of two ways: "inline zomes", or "test wasms".

### How to create "inline zomes"

> The quickest, most preferable way to reproduce a problem is through "inline zomes". Inline zomes are written in terms of a collection of functions, like normal Wasm zomes, but they don't get compiled to wasm, and instead are run inline by Rust directly. This lightweight approach to writing zomes is well suited for quick test cases, or for cases that require a multitude of zomes in order to reproduce a problem.

Reviewer: ```suggestion
The quickest, most preferable way to reproduce a problem is through "inline zomes". Inline zomes are written in terms of a collection of functions, like normal wasm zomes, but they don't get compiled to wasm, and instead are run inline by Rust directly. This lightweight approach to writing zomes is well suited for quick test cases, or for cases that require a multitude of zomes in order to reproduce a problem.
```

> To create a test based on inline zomes, see existing tests using `InlineZomeSet` or `SweetInlineZomes` for guidance. Just put your test in a place that feels appropriate.

Reviewer 2: Feels like it deserves some links.

### How to create a "test wasm"

> Holochain has many "test wasms", which are sample zomes written to demonstrate specific functionality. These

Reviewer: ```suggestion
Holochain has many "test wasms", which are sample zomes written to demonstrate specific functionality.
```

> It may be necessary to write a test wasm if the problem you've encountered has to do with the actual machinery of executing Wasm code. It may also be a good option if you discovered a problem while writing a zome yourself, in which case you can simply copy and paste the offending code into a new test wasm.

Reviewer: ```suggestion
It may be necessary to write a test wasm if the problem you've encountered has to do with the actual machinery of executing wasm code. It may also be a good option if you discovered a problem while writing a zome yourself, in which case you can simply copy and paste the offending code into a new test wasm.
```

Author: The canonical representation that I see is "Wasm", like on [the wikipedia page](https://en.wikipedia.org/wiki/WebAssembly)

Reviewer: Good, I'm fine with any spelling, be it wasm, Wasm or WASM.

To create a test wasm:

1. Create a new crate in `crates/test_utils/wasm/wasm_workspace`
  - See the other test wasms for guidance on proper setup. In particular:
  - The `integrity.lib` will become your integrity zome, and `lib.rs` will become your coordinator zome. Your coordinator zome will be named after the crate name, and the integrity zome will be named with a `_integrity` suffix added.
2. Add your test wasm name to the `TestWasm` enum in [crates/test_utils/wasm/src/lib.rs](https://github.com/holochain/holochain/blob/1b663ec03c86462646cc7693391702a1de02b3a6/crates/test_utils/wasm/src/lib.rs#L21).

>   - Make sure its PascalCase name matches the snake_case name of the crate.

Reviewer 2: Feels like this deserves an example.

>   - You will have to make two other changes in the same file to specify the mapping to snake case, and the location of the built wasm, which will be straightforward to do by observing how other test wasms have done it.

Reviewer 2: link to example?

> 3. Add your test wasm's crate name to the `[workspace]` section of `crates/test_utils/wasm/wasm_workspace/Cargo.toml`.

Reviewer 2: make this path a link that actually links to the real file?

4. Build your test wasm (and all others) with `cargo build --features 'build_wasms' --manifest-path=crates/holochain/Cargo.toml`.

> To write a test using your test wasm, you can use `sweettest::SweetDnaFile::unique_from_test_wasms(vec![TestWasm::YourWasm])` to set up your DNA. See existing tests which use this function to inspiration. [Here is a simple example](https://github.com/holochain/holochain/blob/1a9d85d79a900ad153843a851797f8d46d6ec0e1/crates/holochain/tests/agent_scaling/mod.rs#L101-L134) to follow for guidance.

Reviewer: ```suggestion
To write a test using your test wasm, you can use `sweettest::SweetDnaFile::unique_from_test_wasms(vec![TestWasm::YourWasm])` to set up your DNA. See existing tests which use this function for inspiration. [Here is a simple example](https://github.com/holochain/holochain/blob/1a9d85d79a900ad153843a851797f8d46d6ec0e1/crates/holochain/tests/agent_scaling/mod.rs#L101-L134) to follow for guidance.
```
